control_client/src/main.c

```diff
@@ -96,127 +98,142 @@ int main(int argc, char **argv) {
         }
     }
 
-    int err;
-    err = pad_init(&pad, "127.0.0.1", port);
-    if (err) {
-        fprintf(stderr, "Could not initialize pad server with error: %s\n", strerror(err));
-        exit(EXIT_FAILURE);
-    }
-
-    /* Handle disconnects */
     signal(SIGINT, handle_term);
-
-    err = pad_connect_forever(&pad);
-    if (err) {
-        fprintf(stderr, "Could not connect to pad server with error: %s\n", strerror(err));
-        exit(EXIT_FAILURE);
-    }
-
-    printf("Connection established!\n");
-
-    /* Send messages in a loop */
-
     char key;
     for (;;) {
+        int err;
+        bool pad_connected = true;
 
-        printf("Press key and hit enter: ");
-        key = getc(stdin);
-        if (key != '\n') {
-            while (getc(stdin) != '\n')
-                ;
+        fprintf(stderr, "Waiting for pad...\n");
```

Comment:
`pad_init` should not be in a loop. If `pad_init` fails, it's because socket creation or setting of an option failed.
This isn't something that can be re-tried as a fix, that is only `pad_connect`

Reply:
I put pad_init in the loop to handle disconnections. When the recv times out I close the socket, the inner for loop exists and the socket is reinitialized, if I didn't put pad_init in the loop it would attempt to read a closed socket.
Also if pad_init fails `exit()` gets called, so there would be no retry.

Reply:
Oh I see, when the timeout is up don't close the connection.
In the real system we shouldn't really be purposely closing the connection either; any disconnect will cause an abort sequence if connection is severed for too long (a few seconds). Instead just assume the most recently sent command failed.

Reply:
I see, this was the simplest way I could implement the control_client detecting if the pad_server went down.
I have considered 2 others:
1. the tcp keepalive socket option, the issue is that it would require different flags for different OSs, quite annoying
2. A manual implementation of tcp keepalive where if the server doesn't send anything back you retry a couple times and if there is still no response the server gets considered down

Reply:
Detection that the pad server has gone down will actually happen "for free" just by using TCP. If we send a message, the pad server must send an ACK (TCP protocol ACK, not the same as the ACK in the hybrid communication spec). If it doesn't send the ACK in the TCP timeout window, our `send` call will result in an error and we'll know the connection is down. The case with your implemented timeout is if the pad server has definitely received our message and the connection is stable, but for some reason didn't respond with the success/failure status. This could be because of a logical bug, etc. In this case, for now, I'd like the control box to just consider the command failed. If the pad box isn't responding because of a bug there's really nothing we can do except indicate that failure.

Reply:
Ok weird, I tried doing that in the first place with send errors but I was getting bytes written instead of errors, will try again

Reply:
I believe `0` bytes written is what's considered a disconnect; try terminating the server and sending something from the
client; you should get 0 bytes sent indicating disconnect.

https://www.man7.org/linux/man-pages/man2/send.2.html

Reply:
I still can't make it return 0, the program crashes if I send a request a second time, but that's about it.
I found this in the send man page

> No indication of failure to deliver is implicit in a send().  Locally detected errors are indicated by a return value of -1.

After a quick claude help to decipher what this means:
> "No indication of failure to deliver is implicit in a send()" means:
> - send() doesn't tell you if the data actually reached the other end
> - It only tells you the data was successfully copied to your local TCP send buffer

Recv on the other hand will return an error if there is something wrong with the socket. I'm going to try to use the errno set by the failed recv to differentiate between a timed out server and a crashed server and see where that brings me

Reply:
This achieved the target behavior, I'm confident about the timeout error options but not about the server down ones
```
err = pad_recv(&pad, &act_ack, sizeof(act_ack));
if (err <= 0) {
    if (errno == EAGAIN || errno == EWOULDBLOCK) {
        fprintf(stderr,
                "Server did not acknowledge actuator request with id %d, reconnecting...\n",
                actuator->act_id);
    } else if (errno == ECONNREFUSED || err == ECONNRESET) {
        fprintf(stderr, "SERVER DOWN, I REPEAT, SERVER DOWN\n");
    }
}
```

Reply:
This was a misunderstanding on my part; TCP does actually detect failures when sent messages aren't ACKed, but it's after a timeout duration. The `send()` call also buffers messages to send and doesn't block until they're actually sent on the wire, so you won't know that your `send()` failed until the following receive.

The problem occurs in the scenario where you `send()` successfully and then `recv()` the ACK successfully, but then the remote disconnects. We won't know about the disconnect until the next send option. I think keepalive may be worth looking into in another PR, but your solution here seems fine!